contrib/core-contract-tests/tests/bns/name_register.test.ts

```diff
@@ -0,0 +1,597 @@
+        ["lease-started-at"]: Cl.uint(6),
+      })
+    );
+  });
+
+  it("should fail registering twice", () => {
+    const name = "bob";
+    const merged = new TextEncoder().encode(`${name}.${cases[0].namespace}${cases[0].salt}`);
+    const sha256 = createHash("sha256").update(merged).digest();
+    const ripemd160 = createHash("ripemd160").update(sha256).digest();
+    simnet.callPublicFn("bns", "name-preorder", [Cl.buffer(ripemd160), Cl.uint(2560000)], bob);
+    simnet.callPublicFn(
+      "bns",
+      "name-register",
+      [
+        Cl.bufferFromUtf8(cases[0].namespace),
+        Cl.bufferFromUtf8(name),
+        Cl.bufferFromUtf8(cases[0].salt),
+        Cl.bufferFromUtf8(cases[0].zonefile),
+      ],
+      bob
+    );
+    simnet.callReadOnlyFn("bns", "resolve-principal", [Cl.standardPrincipal(bob)], alice);
+
+    const { result } = simnet.callPublicFn(
+      "bns",
+      "name-register",
+      [
+        Cl.bufferFromUtf8(cases[0].namespace),
+        Cl.bufferFromUtf8(name),
+        Cl.bufferFromUtf8(cases[0].salt),
+        Cl.bufferFromUtf8(cases[0].zonefile),
+      ],
+      bob
+    );
+    expect(result).toBeErr(Cl.int(2004));
+  });
+});
+
+describe("register a name again before and after expiration", () => {
+  beforeEach(() => {
+    // launch namespace
+    const mergedNS = new TextEncoder().encode(`${cases[0].namespace}${cases[0].salt}`);
+    const sha256NS = createHash("sha256").update(mergedNS).digest();
+    const ripemd160NS = createHash("ripemd160").update(sha256NS).digest();
+    simnet.callPublicFn(
+      "bns",
+      "namespace-preorder",
+      [Cl.buffer(ripemd160NS), Cl.uint(cases[0].value)],
+      cases[0].namespaceOwner
+    );
+
+    simnet.callPublicFn(
+      "bns",
+      "namespace-reveal",
+      [
+        Cl.bufferFromUtf8(cases[0].namespace),
+        Cl.bufferFromUtf8(cases[0].salt),
+        ...cases[0].priceFunction,
+        Cl.uint(cases[0].renewalRule),
+        Cl.standardPrincipal(cases[0].nameImporter),
+      ],
+      cases[0].namespaceOwner
+    );
+
+    simnet.callPublicFn(
+      "bns",
+      "namespace-ready",
+      [Cl.bufferFromUtf8(cases[0].namespace)],
+      cases[0].namespaceOwner
+    );
+
+    // register bob.blockstack
+    const name = "bob";
+    const merged = new TextEncoder().encode(`${name}.${cases[0].namespace}${cases[0].salt}`);
+    const sha256 = createHash("sha256").update(merged).digest();
+    const ripemd160 = createHash("ripemd160").update(sha256).digest();
+    simnet.callPublicFn("bns", "name-preorder", [Cl.buffer(ripemd160), Cl.uint(2560000)], bob);
+
+    simnet.callPublicFn(
+      "bns",
+      "name-register",
+      [
+        Cl.bufferFromUtf8(cases[0].namespace),
+        Cl.bufferFromUtf8(name),
+        Cl.bufferFromUtf8(cases[0].salt),
+        Cl.bufferFromUtf8(cases[0].zonefile),
+      ],
+      bob
+    );
+  });
+
+  it("fails if someone else tries to register it", () => {
+    const name = "bob";
+    let salt = "1111";
```

Comment:
This as well may be a good candidate for a fast-check generator (in a future PR).

Reply:
Maybe in a future PR. We could consider it as part of https://github.com/stacks-network/stacks-blockchain/issues/4030